### DISCUSSION TITLE
Wrong zip format or file is corrupted

### DISCUSSION THREAD

**Discussion by forenheith** [icsharpcode/SharpZipLib#561] [Q&A] [answered]
I create a zip file using OutputStream, as result I getting zip file, I can see it's content but when I trying to open it I have the error. What am I doing wrong?
https://gist.github.com/forenheith/910b22ca833c1e3453b32e053d287f9f

**Comment by piksel** [comment, marked as answer]
You would need to provide some code, or else it would be really hard to guess what the problem is. Perhaps you're not disposing the stream? 🤷‍♀️

**Reply by forenheith** [reply]
https://gist.github.com/forenheith/910b22ca833c1e3453b32e053d287f9f

**Reply by piksel** [reply]
I guess I am really good at guessing...

```cs
var zipOutputStream = new ZipOutputStream(outStream);
```

You are not disposing the `ZipOutputStream` and hence, the end of the archive is never written to the output.

```cs
using (var outStream = new MemoryStream())
{
    using(var zipOutputStream = new ZipOutputStream(outStream)) {
        await outStream.FlushAsync(cancellationToken);
        outStream.Seek(0, SeekOrigin.Begin);
        
        // Set IsStreamOwner to false to prevent it from disposing its base stream
        zipOutputStream.IsStreamOwner = false;

        zipOutputStream.SetLevel(CompressionMax);
        foreach (var file in files)
        {
            using (var inStream = new MemoryStream())
            {
                await _fileStorageApi.ReadToStreamAsync(inStream, file.Location.ContainerName,
                    file.Location.FileName, cancellationToken);
                var inBuffer = new byte[inStream.Length];
                inStream.Seek(0, SeekOrigin.Begin);
                var length = await inStream.ReadAsync(inBuffer, 0, inBuffer.Length, cancellationToken);
                var entry = new ZipEntry(file.OriginalName)
                {
                    DateTime = DateTime.Now, Size = inBuffer.Length
                };
                zipOutputStream.PutNextEntry(entry);
                await zipOutputStream.WriteAsync(inBuffer, 0, length, cancellationToken);
            }
        }
    }
    
    outStream.Seek(0, SeekOrigin.Begin);
    var outBuffer = new byte[outStream.Length];
    await outStream.ReadAsync(outBuffer, 0, outBuffer.Length, cancellationToken);
    return outBuffer;
}
```

**Reply by forenheith** [reply]
Why you have decided so? 
                zipOutputStream.Finish();
                zipOutputStream.Close();

**Reply by piksel** [reply]
Yeah, but that is after you have read the output buffer. You need to close it before reading.

**Reply by forenheith** [reply]
> Yeah, but that is after you have read the output buffer. You need to close it before reading.

If I'm disposing zipOutputStream, OutputStream disposes as well, so that I'm getting disposed stream access exception, when I'm trying to read the OutputStream

**Reply by piksel** [reply]
Thats why you need to set the `IsStreamOwner`:
```cs
        // Set IsStreamOwner to false to prevent it from disposing its base stream
        zipOutputStream.IsStreamOwner = false;
```

**Reply by forenheith** [reply]
That works now. Thanks a lot @piksel
